File: src/lib/engine/custom-progressions.ts
// Custom Progressions — parse user input, store/load, provide presets

import { CHORD_INTERVALS, CHORD_NOTATIONS, type NotationStyle } from './chords';
import { noteToSemitone, type AccidentalPreference } from './notes';

// ─── Types ──────────────────────────────────────────────────

export interface CustomChord {
	/** Display text as entered, e.g. "Dm7" */
	display: string;
	/** Root note, e.g. "D" */
	root: string;
	/** Internal quality key, e.g. "m7" */
	quality: string;
	/** How many beats this chord lasts (default 4) */
	beats: number;
}

export interface CustomProgression {
	id: string;
	name: string;
	/** User-entered chord string, e.g. "Dm7 | G7 | CMaj7" */
	raw: string;
	chords: CustomChord[];
	/** BPM for playback */
	bpm: number;
	/** How many times the progression loops (0 = infinite until user stops) */
	loops: number;
	/** Creation timestamp */
	createdAt: number;
}

export interface LoopEvaluation {
	/** Per-chord results for this loop pass */
	chords: ChordEval[];
	/** Overall accuracy (0–1) for this pass */
	accuracy: number;
}

export interface ChordEval {
	/** The chord that was expected */
	chord: CustomChord;
	/** Was the correct chord played at some point during this chord's window? */
	hit: boolean;
	/** Timing offset in ms from the downbeat (negative = early, positive = late) */
	timingOffsetMs: number;
	/** Best accuracy achieved (0–1) during the window */
	bestAccuracy: number;
}

export interface SessionEvaluation {
	/** Results per loop pass */
	loops: LoopEvaluation[];
	/** Overall hit rate */
	overallAccuracy: number;
	/** Average absolute timing offset */
	avgTimingMs: number;
	/** Chords that were consistently missed */
	weakChords: string[];
	/** Total elapsed ms */
	totalMs: number;
}

// ─── Chord Parsing ──────────────────────────────────────────

/**
 * All quality labels we recognize — longest first so "Maj7#11" matches before "Maj7".
 * Built from CHORD_NOTATIONS + some common aliases.
 */
function buildQualityMap(): Map<string, string> {
	const map = new Map<string, string>();
	// Add all notation variants (longest match first)
	for (const style of Object.keys(CHORD_NOTATIONS) as NotationStyle[]) {
		for (const [internal, display] of Object.entries(CHORD_NOTATIONS[style])) {
			map.set(display, internal);
		}
	}
	// Common aliases people might type
	map.set('Δ7', 'Maj7');
	map.set('Δ9', 'Maj9');
	map.set('-7', 'm7');
	map.set('-9', 'm9');
	map.set('-11', 'm11');
	map.set('ø7', 'm7b5');
	map.set('ø', 'm7b5');
	map.set('dim7', 'dim7');
	map.set('dim', 'dim7');
	map.set('o7', 'dim7');
	map.set('maj7', 'Maj7');
	map.set('maj9', 'Maj9');
	map.set('min7', 'm7');
	map.set('min9', 'm9');
	map.set('M7', 'Maj7');
	map.set('M9', 'Maj9');
	map.set('mi7', 'm7');
	map.set('mi9', 'm9');
	return map;
}

const qualityMap = buildQualityMap();
/** Quality strings sorted longest-first for greedy matching */
const qualityKeys = [...qualityMap.keys()].sort((a, b) => b.length - a.length);

/**
 * Parse a single chord symbol like "Dm7", "BbMaj7", "F#7#9".
 * Returns null if unrecognized.
 */
export function parseChordSymbol(input: string): { root: string; quality: string; display: string } | null {
	const trimmed = input.trim();
	if (!trimmed) return null;

	// Extract root: A-G followed by optional # or b
	const rootMatch = trimmed.match(/^([A-G][#b♯♭]?)/);
	if (!rootMatch) return null;

	let root = rootMatch[1].replace('♯', '#').replace('♭', 'b');
	const rest = trimmed.slice(rootMatch[0].length);

	// If nothing follows, it's probably a major triad — not in our system
	if (!rest) return null;

	// Try to match quality from longest to shortest
	// First pass: exact match only (preserves case sensitivity: M7≠m7)
	for (const key of qualityKeys) {
		if (rest === key) {
			const quality = qualityMap.get(key)!;
			if (CHORD_INTERVALS[quality]) {
				return { root, quality, display: trimmed };
			}
		}
	}
	// Second pass: case-insensitive fallback for user convenience
	for (const key of qualityKeys) {
		if (rest.toLowerCase() === key.toLowerCase()) {
			const quality = qualityMap.get(key)!;
			// Verify we have intervals for this quality
			if (CHORD_INTERVALS[quality]) {
				return { root, quality, display: trimmed };
			}
		}
	}

	// Direct lookup in CHORD_INTERVALS (user typed internal key like "Maj7")
	if (CHORD_INTERVALS[rest]) {
		return { root, quality: rest, display: trimmed };
	}

	return null;
}

/**
 * Parse a full progression string.
 * Accepts: "Dm7 | G7 | CMaj7" or "Dm7 - G7 - CMaj7" or "Dm7 G7 CMaj7"
 * Also accepts beat annotations: "Dm7(2) | G7(2) | CMaj7(4)"
 */
export function parseProgression(input: string): CustomChord[] {
	// Normalize separators: | or – or - or , → consistent delimiter
	const normalized = input
		.replace(/[|–\-,]/g, ' ')
		.replace(/\s+/g, ' ')
		.trim();

	if (!normalized) return [];

	const tokens = normalized.split(' ').filter(Boolean);
	const chords: CustomChord[] = [];

	for (const token of tokens) {
		// Check for beat annotation: "Dm7(2)"
		const beatMatch = token.match(/^(.+)\((\d+)\)$/);
		const chordStr = beatMatch ? beatMatch[1] : token;
		const beats = beatMatch ? parseInt(beatMatch[2], 10) : 4;

		const parsed = parseChordSymbol(chordStr);
		if (parsed) {
			chords.push({
				display: parsed.display,
				root: parsed.root,
				quality: parsed.quality,
				beats,
			});
		}
	}

	return chords;
}

/**
 * Format a progression back into a display string.
 */
export function formatProgression(chords: CustomChord[]): string {
	return chords.map((c) => {
		if (c.beats !== 4) return `${c.display}(${c.beats})`;
		return c.display;
	}).join(' | ');
}

// ─── Preset Progressions ────────────────────────────────────

export interface ProgressionPreset {
	name: string;
	/** Genre/source */
	tag: string;
	raw: string;
	bpm: number;
}

export const PROGRESSION_PRESETS: ProgressionPreset[] = [
	{
		name: 'Autumn Leaves (A-Teil)',
		tag: 'Jazz Standard',
		raw: 'Cm7 | F7 | BbMaj7 | EbMaj7 | Am7b5 | D7 | Gm7 | Gm7',
		bpm: 140,
	},
	{
		name: 'All of Me (A-Teil)',
		tag: 'Jazz Standard',
		raw: 'CMaj7 | CMaj7 | E7 | E7 | A7 | A7 | Dm7 | Dm7',
		bpm: 130,
	},
	{
		name: 'Blue Bossa',
		tag: 'Latin Jazz',
		raw: 'Cm7 | Cm7 | Fm7 | Fm7 | Dm7b5 | G7 | Cm7 | Cm7',
		bpm: 140,
	},
	{
		name: 'Rhythm Changes (A)',
		tag: 'Bebop',
		raw: 'BbMaj7 | Gm7 | Cm7 | F7 | Dm7 | Gm7 | Cm7 | F7',
		bpm: 160,
	},
	{
		name: '12-Bar Blues',
		tag: 'Blues',
		raw: 'C7 | C7 | C7 | C7 | F7 | F7 | C7 | C7 | G7 | F7 | C7 | G7',
		bpm: 120,
	},
	{
		name: 'So What',
		tag: 'Modal Jazz',
		raw: 'Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7',
		bpm: 136,
	},
	{
		name: 'Satin Doll (A)',
		tag: 'Swing',
		raw: 'Dm7 | G7 | Em7 | A7 | Am7 | D7 | AbMaj7 | DbMaj7',
		bpm: 120,
	},
];

// ─── Storage ────────────────────────────────────────────────

const STORAGE_KEY = 'chord-trainer-custom-progressions';
const MAX_SAVED = 50;

function generateId(): string {
	return Date.now().toString(36) + Math.random().toString(36).slice(2, 6);
}

export function loadCustomProgressions(): CustomProgression[] {
	if (typeof localStorage === 'undefined') return [];
	try {
		const raw = localStorage.getItem(STORAGE_KEY);
		return raw ? JSON.parse(raw) : [];
	} catch {
		return [];
	}
}

export function saveCustomProgression(prog: Omit<CustomProgression, 'id' | 'createdAt'>): CustomProgression {
	const full: CustomProgression = {
		...prog,
		id: generateId(),
		createdAt: Date.now(),
	};
	const all = loadCustomProgressions();
	all.unshift(full);
	if (all.length > MAX_SAVED) all.length = MAX_SAVED;
	if (typeof localStorage !== 'undefined') {
		localStorage.setItem(STORAGE_KEY, JSON.stringify(all));
	}
	return full;
}

export function deleteCustomProgression(id: string): void {
	const all = loadCustomProgressions().filter((p) => p.id !== id);
	if (typeof localStorage !== 'undefined') {
		localStorage.setItem(STORAGE_KEY, JSON.stringify(all));
	}
}

// ─── Evaluation ─────────────────────────────────────────────

/**
 * Given an array of per-chord evaluations from one or more loop passes,
 * compute the overall session evaluation.
 */
export function evaluateSession(loops: LoopEvaluation[], totalMs: number): SessionEvaluation {
	const allChordEvals = loops.flatMap((l) => l.chords);
	const hitCount = allChordEvals.filter((e) => e.hit).length;
	const totalCount = allChordEvals.length;
	const avgTiming = totalCount > 0
		? allChordEvals.reduce((sum, e) => sum + Math.abs(e.timingOffsetMs), 0) / totalCount
		: 0;

	// Track miss count per chord display
	const missCounts = new Map<string, number>();
	for (const e of allChordEvals) {
		if (!e.hit) {
			missCounts.set(e.chord.display, (missCounts.get(e.chord.display) ?? 0) + 1);
		}
	}

	// Weak = missed more than half the time
	const loopCount = loops.length || 1;
	const weakChords = [...missCounts.entries()]
		.filter(([, count]) => count > loopCount / 2)
		.map(([chord]) => chord);

	return {
		loops,
		overallAccuracy: totalCount > 0 ? hitCount / totalCount : 0,
		avgTimingMs: Math.round(avgTiming),
		weakChords,
		totalMs,
	};
}

File: ios/MusicEngine/Sources/MusicEngine/CustomProgressions.swift
// Custom Progressions — parse user input, presets, evaluation.
// Port of the PURE parts of src/lib/engine/custom-progressions.ts.
// Storage (localStorage) is handled by the app's persistence layer, not here.

import Foundation

public struct CustomChord: Sendable, Equatable, Codable {
    public let display: String
    public let root: String
    public let quality: String
    public let beats: Int

    public init(display: String, root: String, quality: String, beats: Int) {
        self.display = display
        self.root = root
        self.quality = quality
        self.beats = beats
    }
}

public struct ParsedChord: Sendable, Equatable {
    public let root: String
    public let quality: String
    public let display: String
}

// ─── Quality map (longest-first greedy matching) ──────────────

private let qualityMap: [String: String] = {
    var map: [String: String] = [:]
    for (_, table) in CHORD_NOTATIONS {
        for (internalKey, display) in table {
            map[display] = internalKey
        }
    }
    // Common aliases people might type
    let aliases: [String: String] = [
        "Δ7": "Maj7", "Δ9": "Maj9", "-7": "m7", "-9": "m9", "-11": "m11",
        "ø7": "m7b5", "ø": "m7b5", "dim7": "dim7", "dim": "dim7", "o7": "dim7",
        "maj7": "Maj7", "maj9": "Maj9", "min7": "m7", "min9": "m9",
        "M7": "Maj7", "M9": "Maj9", "mi7": "m7", "mi9": "m9",
    ]
    for (k, v) in aliases { map[k] = v }
    return map
}()

private let qualityKeys: [String] = qualityMap.keys.sorted { $0.count > $1.count }

/// Parse a single chord symbol like "Dm7", "BbMaj7", "F#7#9". Returns nil if unrecognized.
public func parseChordSymbol(_ input: String) -> ParsedChord? {
    let trimmed = input.trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty { return nil }

    // Extract root: A-G followed by optional accidental
    guard let rootRange = trimmed.range(of: "^[A-G][#b♯♭]?", options: .regularExpression) else {
        return nil
    }
    var root = String(trimmed[rootRange])
    root = root.replacingOccurrences(of: "♯", with: "#").replacingOccurrences(of: "♭", with: "b")
    let rest = String(trimmed[rootRange.upperBound...])

    if rest.isEmpty { return nil }

    // First pass: exact match (case-sensitive: M7 ≠ m7)
    for key in qualityKeys where rest == key {
        if let quality = qualityMap[key], CHORD_INTERVALS[quality] != nil {
            return ParsedChord(root: root, quality: quality, display: trimmed)
        }
    }
    // Second pass: case-insensitive fallback
    for key in qualityKeys where rest.lowercased() == key.lowercased() {
        if let quality = qualityMap[key], CHORD_INTERVALS[quality] != nil {
            return ParsedChord(root: root, quality: quality, display: trimmed)
        }
    }
    // Direct lookup (user typed internal key like "Maj7")
    if CHORD_INTERVALS[rest] != nil {
        return ParsedChord(root: root, quality: rest, display: trimmed)
    }

    return nil
}

/// Parse a full progression string. Accepts "Dm7 | G7 | CMaj7", dashes, commas,
/// and beat annotations "Dm7(2)".
public func parseProgression(_ input: String) -> [CustomChord] {
    // Normalize separators: | – - , → space
    var normalized = input
    for sep in ["|", "–", "-", ","] {
        normalized = normalized.replacingOccurrences(of: sep, with: " ")
    }
    normalized = normalized.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        .trimmingCharacters(in: .whitespaces)

    if normalized.isEmpty { return [] }

    let tokens = normalized.split(separator: " ").map(String.init).filter { !$0.isEmpty }
    var chords: [CustomChord] = []

    for token in tokens {
        var chordStr = token
        var beats = 4
        // Beat annotation: "Dm7(2)"
        if let m = token.range(of: "^(.+)\\((\\d+)\\)$", options: .regularExpression) {
            let matched = String(token[m])
            if let paren = matched.firstIndex(of: "("), let close = matched.firstIndex(of: ")") {
                chordStr = String(matched[matched.startIndex..<paren])
                let beatStr = String(matched[matched.index(after: paren)..<close])
                beats = Int(beatStr) ?? 4
            }
        }

        if let parsed = parseChordSymbol(chordStr) {
            chords.append(CustomChord(display: parsed.display, root: parsed.root, quality: parsed.quality, beats: beats))
        }
    }

    return chords
}

/// Format a progression back into a display string.
public func formatProgression(_ chords: [CustomChord]) -> String {
    chords.map { c in
        c.beats != 4 ? "\(c.display)(\(c.beats))" : c.display
    }.joined(separator: " | ")
}

// ─── Preset Progressions ────────────────────────────────────

public struct ProgressionPreset: Sendable, Equatable {
    public let name: String
    public let tag: String
    public let raw: String
    public let bpm: Int
}

public let PROGRESSION_PRESETS: [ProgressionPreset] = [
    ProgressionPreset(name: "Autumn Leaves (A-Teil)", tag: "Jazz Standard", raw: "Cm7 | F7 | BbMaj7 | EbMaj7 | Am7b5 | D7 | Gm7 | Gm7", bpm: 140),
    ProgressionPreset(name: "All of Me (A-Teil)", tag: "Jazz Standard", raw: "CMaj7 | CMaj7 | E7 | E7 | A7 | A7 | Dm7 | Dm7", bpm: 130),
    ProgressionPreset(name: "Blue Bossa", tag: "Latin Jazz", raw: "Cm7 | Cm7 | Fm7 | Fm7 | Dm7b5 | G7 | Cm7 | Cm7", bpm: 140),
    ProgressionPreset(name: "Rhythm Changes (A)", tag: "Bebop", raw: "BbMaj7 | Gm7 | Cm7 | F7 | Dm7 | Gm7 | Cm7 | F7", bpm: 160),
    ProgressionPreset(name: "12-Bar Blues", tag: "Blues", raw: "C7 | C7 | C7 | C7 | F7 | F7 | C7 | C7 | G7 | F7 | C7 | G7", bpm: 120),
    ProgressionPreset(name: "So What", tag: "Modal Jazz", raw: "Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Ebm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7 | Dm7", bpm: 136),
    ProgressionPreset(name: "Satin Doll (A)", tag: "Swing", raw: "Dm7 | G7 | Em7 | A7 | Am7 | D7 | AbMaj7 | DbMaj7", bpm: 120),
]

// ─── Evaluation ─────────────────────────────────────────────

public struct ChordEval: Sendable, Equatable {
    public let chord: CustomChord
    public let hit: Bool
    public let timingOffsetMs: Double
    public let bestAccuracy: Double

    public init(chord: CustomChord, hit: Bool, timingOffsetMs: Double, bestAccuracy: Double) {
        self.chord = chord
        self.hit = hit
        self.timingOffsetMs = timingOffsetMs
        self.bestAccuracy = bestAccuracy
    }
}

public struct LoopEvaluation: Sendable, Equatable {
    public let chords: [ChordEval]
    public let accuracy: Double

    public init(chords: [ChordEval], accuracy: Double) {
        self.chords = chords
        self.accuracy = accuracy
    }
}

public struct SessionEvaluation: Sendable, Equatable {
    public let loops: [LoopEvaluation]
    public let overallAccuracy: Double
    public let avgTimingMs: Int
    public let weakChords: [String]
    public let totalMs: Double
}

/// Compute the overall session evaluation from per-loop results.
public func evaluateSession(_ loops: [LoopEvaluation], _ totalMs: Double) -> SessionEvaluation {
    let allChordEvals = loops.flatMap { $0.chords }
    let hitCount = allChordEvals.filter { $0.hit }.count
    let totalCount = allChordEvals.count
    let avgTiming = totalCount > 0
        ? allChordEvals.reduce(0.0) { $0 + abs($1.timingOffsetMs) } / Double(totalCount)
        : 0.0

    var missCounts: [String: Int] = [:]
    for e in allChordEvals where !e.hit {
        missCounts[e.chord.display, default: 0] += 1
    }

    let loopCount = loops.isEmpty ? 1 : loops.count
    let weakChords = missCounts
        .filter { Double($0.value) > Double(loopCount) / 2 }
        .map { $0.key }

    return SessionEvaluation(
        loops: loops,
        overallAccuracy: totalCount > 0 ? Double(hitCount) / Double(totalCount) : 0,
        avgTimingMs: Int(avgTiming.rounded()),
        weakChords: weakChords,
        totalMs: totalMs
    )
}
